### ISSUE TITLE
[ACP]  RangeBounds::overlaps

### ISSUE THREAD

**Issue by theli-ua** [rust-lang/libs-team#277]
# Proposal

## Problem statement
RangeBounds today is a standard trait to represent ranges and  provides a single method `contains` used to determine if a range contains a given element.
Another common operation on ranges (like intervals) is to check if 2 overlap. Today there is no method to check for that

## Motivating examples or use cases
* a database implementation using btree index where a parent node would split its total key range in ranges per child. To implement an operation to find keys within a given range it would need to check if child ranges overlap with an input range
* An interval map implementation would need to be able to check for overlaps
* A memory copy/move implementation operation may check for overlaps to decide if it needs to use move or copy
* A database implementation allowing to operate on ranges (eg range deletes, range queries, etc) would need to have an ability to check for range overlaps

## Solution sketch
```
pub trait RangeBounds<T: ?Sized> {

    /// Returns `true` if there exists an element present in both ranges.
    ///
    /// # Examples
    ///
    /// ```
    /// assert!( (3..5).overlaps(&(1..4));
    /// assert!(!(3..5).overlaps(&(5..6));
    /// ```
    ///
    fn overlaps<O, E>(&self, other: &O) -> bool
    where
        T: PartialOrd<E>,
        E: ?Sized + PartialOrd<T>,
        O: RangeBounds<E>,
    {
          todo!()
    }
}


```
## Open questions

There is a question whether to consider  (Excluded(1), Excluded(3)) as overlapping with (Excluded(2), Excluded(5)).
Since this is only a problem for cases where a discrete step between elements is known we could have a specialized implementation for anything that implements `Step` that handles these cases.

## Alternatives
Alternative would be for developers to implement it on specific types, or an extension trait on top of RangeBounds (which *could* be a separate crate)

**Comment by pitaj** [comment]
In most cases, don't people want to perform some sort of operation on the overlapping region? Could the function return a range instead? Then you could use `Range::is_empty` to check for an actual overlap.

**Comment by theli-ua** [comment]
@pitaj well, we'd probably need to add `is_empty` to `RangeBounds` in that case.

**Comment by theli-ua** [comment]
Yeah, given that this is a trait and in case of intersect/overlap we would need to return a concrete type I'd say for the trait specifically it makes sense to have `overlaps` that returns a boolean and would be similar to currently provided method - `contains`

**Comment by quaternic** [comment]
```rust
  fn overlaps<O, E>(&self, other: &O) -> bool
  where
      T: PartialOrd<E>,
      E: ?Sized + PartialOrd<T>,
      O: RangeBounds<E>,
```

Functions like `overlaps` or `intersect` should require `T: Ord`, because if the upper (or lower) bounds of the two sets are not mutually comparable, you can't necessarily compute the intersection as a range. It might not even be representable as a range at all.

Example: Nonnegative integers can be partially ordered by divisibility, so that e.g. `D(2)..=D(30)`  contains `D(x)` iff `x` is even and divides `30`. Then the intersection of the ranges `D(2)..=D(30)` and `D(3)..=D(12)` is `D(6)..=D(6)`, but computing that requires more than what `PartialOrd` provides.

**Comment by theli-ua** [comment]
@quaternic that makes sense so it'd be
```
pub trait RangeBounds<T: ?Sized> {

    /// Returns `true` if there exists an element present in both ranges.
    ///
    /// # Examples
    ///
    /// ```
    /// assert!( (3..5).overlaps(&(1..4));
    /// assert!(!(3..5).overlaps(&(5..6));
    /// ```
    ///
    fn overlaps(&self, other: &T) -> bool
    where
        T: Ord,
    {
          todo!()
    }
}

```

**Comment by pitaj** [comment]
> Hmm, `intersect` probably wants different return types depending on the input ranges, not just an `impl RangeBounds`.
> 
> Is it reasonable to mix the types?  I guess `Range & RangeInclusive` doesn't know the type to return, so maybe not?

> Yeah, given that this is a trait and in case of intersect/overlap we would need to return a concrete type I'd say for the trait specifically it makes sense to have overlaps that returns a boolean

I think we can just return `(Bound, Bound)` rather than a concrete range type to avoid this issue.

**Comment by quaternic** [comment]
`Bound` on its own doesn't have a meaningful (unique) ordering, since it depends on whether it is used as the upper or lower bound:
As a lower bound, `Unbounded < ... < Included(1) < Excluded(1) < Included(2) < Excluded(2) < ...`
As an upper bound, `... < Excluded(1) < Included(1) < Excluded(2) < Included(2) < ... < Unbounded`

But taking that into account by defining `max_lower` and `min_upper`, I agree that
> the right behaviour would be `Range::intersect(a..b, c..d) => max_lower(a, c)..min_upper(b, d)`

where it turns out that both `max_lower(a,c)` and `min_upper(b,d)` have semantics of "choose according to the respective operator on the inner values, prefer Excluded when inner values are equal, and only Unbounded if both are".

**Comment by quaternic** [comment]
Yes, the minimum of two lower bounds (or the maximum of two upper bounds) would prefer `Included` over `Excluded` on equal inner values. These would similarly be used for the operation of computing the least range that contains both input ranges. E.g. `(2..13).join(11..13) == 2..13` and `(2..13).join(11..=13) == 2..=13`. (That one is probably harder to justify for the standard range types than `intersect`, but comes up at least for bounding boxes.)

We won't be changing it now (or ever), but for the sake of discussion one could consider an alternative bound-type defined as
```
enum Bound<T> {
    BeforeAny,
    Before(T),
    After(T),
    AfterAny,
}
```
This wouldn't suffer from the same ambiguity, as you'd just define `BeforeAny < ... < Before(0) < 0 < After(0) < Before(1) < 1 < After(1) < ... AfterAny`

And the ranges would be like
 - 0..=5 -> `(Before(0), After(5))`
 - 0..5 -> `(Before(0), Before(5))`
 - 0.. -> `(Before(0), AfterAny)`
 - .. -> `(BeforeAny, AfterAny)`

**Comment by pitaj** [comment]
I've implemented `clamp` (intersection) in [this playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=a4859e32425aa6445c11f90e66b7f972) along with `is_empty` and `into_bounds`. It has to work with the bounds by value, otherwise there's ambiguity (multiple applicable impls) as seen [here](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f4611e8b67986c626233b3ed9cf46feb)

**Comment by kennytm** [comment]
@pitaj thanks for the hard work

But if you need to go through that `LessNonAdjacent` mess to get `(Excluded(1_u8), Excluded(2_u8))` return is-empty, I have doubt why `..0_u8` (`(Unbounded, Excluded(0_u8))`) remains treated as not-empty.

**Comment by quaternic** [comment]
Indeed.`(Exluded(1), Excluded(2))` should  be considered non-empty as a range. Sure, there isn't an `x: u8` that would satisfy `1_u8 < x < 2_u8`, but conceptually the range represents that predicate rather than any concrete set of values.

In fact, the `contains`-method allows for any type `U: PartialOrd<T>`, so you can implement another type that does have values in that range: [playground](https://play.rust-lang.org/?gist=ba1ebb821dc6cfe1f3dbd89c2f62fd29)
```
[src/main.rs:39:5] (Excluded(1), Excluded(2)).contains(&IntPlusHalf(1)) = true
```
On the other hand, when `a >= b`, the emptiness of `a..b` is implied by the required properties of `PartialOrd`.

**Comment by pitaj** [comment]
Great point! I've removed that and now `(Excluded(1), Excluded(2))` is considered to be non-empty.

[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=54837d7048f3fbcaab9aba77bde67fbe)

**Comment by judnich** [comment]
> Indeed.`(Exluded(1), Excluded(2))` should be considered non-empty as a range. Sure, there isn't an `x: u8` that would satisfy `1_u8 < x < 2_u8`, but conceptually the range represents that predicate rather than any concrete set of values.
> 
> In fact, the `contains`-method allows for any type `U: PartialOrd<T>`, so you can implement another type that does have values in that range: [playground](https://play.rust-lang.org/?gist=ba1ebb821dc6cfe1f3dbd89c2f62fd29)
> 
> ```
> [src/main.rs:39:5] (Excluded(1), Excluded(2)).contains(&IntPlusHalf(1)) = true
> ```
> 
> On the other hand, when `a >= b`, the emptiness of `a..b` is implied by the required properties of `PartialOrd`.

### This is dangerously incorrect for generic T

I really can't stress enough how dangerous this interpretation or definition of "emptiness" is. It has and will continue to cause dangerous and subtle bugs in downstream code that uses this `is_empty()` function.

I argue that virtually everyone will naturally interpret `is_empty() -> bool` to be defined as `{ the set of items this interval matches } == {}`. This is how intervals are defined mathematically, and it is indeed the most logically consistent and what most people will be expecting. It is also deceptive to see `is_empty() -> bool` because it suggests that it's possible to determine whether this set is empty or not, when in fact this function really should be returning `Option<bool>`.

_It is mathematically impossible to determine whether a double open interval is empty or not without knowing more about the domain in which the bounds live,_ or at least some key properties of that space (namely, we need a successor or predecessor function defined on `T`).

* The double-open interval `(1, 2)` IS EMPTY for integer T.
* The double-open interval `(1, 2)` is NOT EMPTY for real T.

### More generally:

* A double-open interval `(a, a+epsilon)` where `epsilon > 0` is EMPTY for discrete T.
* A double-open interval `(a, b+epsilon)` where `epsilon > 0` is NON EMPTY for continuous T.

### In general:

* It is theoretically IMPOSSIBLE to decide non-emptiness of an arbitrary double-open using comparison operators alone.
* As such (with no extra traits on T available), it is only possible to be correct if we define `is_empty() -> Option<bool>`
* Otherwise, to establish non-emptiness of a double-open interval, we need extra knowledge about T.

### Extra knowledge about T required:

* For maximally general/arbitrary `T`, to correctly decide boolean emptiness, a function must exist for `T` that decides whether one or more elements exist within the double-open interval `(a, b)`.
* A successor function (`(x ∈ T)` ↦ smallest `y ∈ T` such that `y > x`) is sufficient though not necessary to derive this function.
* For some `T` domains, this function can be defined without knowing the specific values of `a` and `b`, merely that they are non-equal. This is true for reals and arbitrary strings, for example, where knowing that `a < b` (without needing any more information about the actual values) is sufficient to conclude that the double-open interval `(a, b)` matches one or more values of `T`.
* However, this is NOT the case for ANY finite `T`! And all fixed-size types are in fact finite, _including floats_ (though in the case of floats we often consider it tolerable to make an exception and treat them as reals in cases like this).

## Recommendation:

Aside from coming up with a complex mechanism to define a successor function or various other set-theoretic approaches here to generically solve `is_empty() -> bool` for arbitrary `RangeBound<T>`, please just define `is_empty() -> Option<bool>` and return `None` when we're operating on `(Excluded(_), Excluded(_))` and for all `PartialOrd` comparisons for which `partial_cmp` returns `None`. This is the only correct way to define and implement `is_empty` in the absence of both (1) total ordering, and (2) a defined successor function or other way to determine non-emptiness of a double-open interval.

**Comment by quaternic** [comment]
> ### Definition of "empty"
> 
> I claim that `range.is_empty() == true` should imply _"there exist no values x ∈ T for which range.contains(x)"._

I believe this is the root disagreement here. Specifically the part "no values x **in T**".

As you've argued, it is not possible to implement `is_empty() -> bool` with that definition.

My point was that `contains` takes values of any type `U` that are comparable with `T`, and it *is* possible (and easy) to define `range.is_empty()` equivalent to "there exist no values `x` *of any type* s.t. `range.contains(&x)`".

So, could you clarify why you find that definition dangerous?

**Comment by judnich** [comment]
Almost any nontrivial algorithm written on top of `RangeBounds<T>` will develop critical bugs when both bounds are `Bound::Excluded`.

Here's a very simple example: Implement a binary search algorithm on top of `RangeBounds<T>` that intentionally excludes the middle/pivot point that was checked from further iterations. After a few iterations (at least one up-step and one down-step), you'll end up with a `RangeBounds<T>` that is `Bound::Excluded` on both ends.

If the termination case for this binary search algorithm is `search_range.is_empty()`, **the binary search will never terminate and hang indefinitely**, despite the code being obviously correctly written using any sane definition of `is_empty()`.

It's very reasonable to write code like this, and this could cause all sorts of nasty subtle bugs in any sort of search, tree, or other code that uses ranges that can be either open or closed on either end.

You might argue "well just use `Range<T>` so that it's inclusive on the lower bound". But no, if I could do that, then I wouldn't be using `RangeBounds<T>` or pairs of `Bounds<T>` in the first place; I want to be able to write algorithms with `RangeBounds` over at least reasonable common types `T` (ints, floats, strings, etc.) without it breaking catastrophically due to edge cases unsoundly defined `is_empty() -> bool`.

This **IS** a real and serious problem, and there's really only two possible solutions:

1. Define `is_empty() -> Option<bool>` instead of `-> bool`.
2. Define some extra trait that allows `is_empty() -> bool` to be implemented correctly and soundly.

I argue that the current approach (define `is_empty() -> bool` unsoundly and implement it incorrectly) is far worse than options (1) or (2).

**Comment by judnich** [comment]
> The documentation currently starts with
> 
> > Returns true if the range contains no items.
> 
> Maybe this would be more accurately expressed as
> 
> > Returns true if the range cannot contain any items.

The latter documentation would be better, but I'd push strongly for renaming the function in that case.

Again, 99.999% of people when seeing `is_empty() -> bool` will assume this returns whether or not (bool) something is empty (is_empty)! This should be pretty clear and obvious to anyone.

If it was instead named `cannot_contain_items` (or something else like that), I don't think anyone would have this problem. `is_empty() -> bool` has a long and well-entrenched meaning, and it most definitely does not mean that.

**Comment by judnich** [comment]
I would argue that just documentation is insufficient to resolve this. I could just as well then argue that we should replace the implementation of `is_empty() -> bool` to return the opposite bool that it does now, and simply add a line into the docs to document this, then we're all good :)

For a function with such a fundamentally simple and clearly understood common language meaning like `is_empty -> true OR false`, I argue you can't just document-away edge case behavior.

Sure, people do this type of thing all the time in various languages an libraries, but I don't think this is good enough for what is expected of the standard library of Rust, a language and runtime that's come to be relied upon for being rigorous and meticulous in the correctness of its standard APIs, *especially* for how edge-cases are handled explicitly whereas other languages/libraries might just brush it off and leave the user to deal with the fallout.

**Comment by quaternic** [comment]
> If the termination case for this binary search algorithm is `search_range.is_empty()`, **the binary search will never terminate and hang indefinitely**, despite the code being obviously correctly written using any sane definition of `is_empty()`.

How does the algorithm pick a new value within the range?

Maybe something like this:
```
while !range.is_empty() {
    // IMO, the bug is here: `x` could be outside `range`
    let x = range.start.0.midpoint(range.end.0);

    match f(x) {
        Less => range.end = Excluded(x),
        Equal => return x,
        Greater => range.start = Excluded(x),
    }
}
```
`range.is_empty() == false` does not imply that any `x` you
actually pick is within the range. Even with your definition,
the above code could hang on `(Excluded(1), Included(2))`.

**Comment by pitaj** [comment]
I'd like to see an actual code example demonstrating the issue.

**Comment by Fabi0San** [comment]
Thanks @judnich for the interesting points.
I agree that the Range<T> api is less than optimal and leaves some gaps.
While writing boolean algebraic simplification on a query planner for a custom database engine I was faced with some of these issues and would like to offer my approach in case it is helpful in your endeavor to get this raised and fixed.

First of all no boundary is really at a value, it is either before or after the value. I encoded that on my Boundary<T>.is_after.
std implementation encodes this info in the Range<T> instead, but makes for a messy implementation as an inclusive start boundary is before the start value but and inclusive end boundary is after the value.

The second trick was creating a trait Adjacency<T>, which should answer the question is_adjacent for any two values.
1 & 2 are adjacent for datatype integer, but not adjacent for datatype real.

From there the determination of is_empy is trivial, ascending, and adjacent values with boudaries placed after and before respectively are considered empty. (1,2)

computing culture aware adjacency of strings is less than trivial, but implementing as a constant false is mostly safe for our uses.

I hope it helps, keep up the good fight!

**Comment by judnich** [comment]
Yeah, I think I've said all that needs to be said -- beyond this, it's up to the judgement of whoever is in charge of the ACP review/design process to take feedback reporting substantive issues with the design seriously, or not.

When the consequences of this design decision (a mistake IMO) are faced [*] if the current nightly API remains, these comments here will at least serve as a record that those in charge were alerted to this issue in great detail, and yet they chose to proceed down the problematic path despite many strictly better alternates being proposed.

[*] Consequences: Whether it's that you'll get `empty.is_empty() == false` for some _empty_ ranges, or `x.overlaps(disjoint_with_x) == true` for some _disjoint_ ranges, or the mathematical equivalent `!x.intersect(disjoint_with_x).is_empty() == true`, or many other such consequences of this current design.
